lib/administrate/search.rb

```diff
@@ -69,11 +69,13 @@ def tables_to_join
     def association_search?(attribute)
       return unless attribute_types[attribute].respond_to?(:deferred_class)
 
-      [
-        Administrate::Field::BelongsTo,
-        Administrate::Field::HasMany,
-        Administrate::Field::HasOne,
-      ].include?(attribute_types[attribute].deferred_class)
+      association_classes.include?(attribute_types[attribute].deferred_class)
+    end
+
+    def association_classes
+      @association_classes ||=
+        ObjectSpace.each_object(Class).
+          select { |klass| klass < Administrate::Field::Associative }
```

Comment:
As an alternative to mapping over `ObjectSpace` (and maybe even skipping the memoization), have you considered using [`Class#descendants`](https://api.rubyonrails.org/classes/Class.html#method-i-descendants)?

```ruby
class C; end
C.descendants # => []

class B < C; end
C.descendants # => [B]

class A < B; end
C.descendants # => [B, A]

class D < C; end
C.descendants # => [B, A, D]
```

Reply:
Could also use `ancestors` from the `deferred_class` to see if it includes `Administrate::Field::Associative`. This would allow for any plugins that extend the `BelongTo`, `HasMany` etc to also be included.

This could be:
```
  def association_search?(attribute)
      return unless attribute_types[attribute].respond_to?(:deferred_class)
      attribute_types[attribute].deferred_class.ancestors.include?(Administrate::Field::Associative)
  end
```

Reply:
In defence of this solution, I'd say it just does the same thing that already happens at [`lib/administrate/base_dashboard.rb`](https://github.com/thoughtbot/administrate/blob/acada0585b35aee532116fe04d65dc5263bba179/lib/administrate/base_dashboard.rb#L76) (and I myself introduced at https://github.com/thoughtbot/administrate/pull/1176/files).

I'd be happy with merging this, then having another PR to refactor both instances.